text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,263 @@
+nav_order: 3
+---
+
+- Feature Name: OSN_snapshots
+- Start Date: 2021-03-23
+- RFC PR:
+- Fabric Component: orderer, ledger
+- Fabric Issue:
+
+# Summary
+[summary]: #summary
+
+The RFC introduces ledger snapshot feature at orderers, similar to the same feature at peers.
+
+# Motivation
+[motivation]: #motivation
+
+Real world use cases show the necessity of controlling the disk space consumed by block ledger, and this RFC is mainly motivated by the need to reduce space consumption at orderer nodes.
+Additionally this RFC presents a better way to join a channel without pulling the huge number of all historical blocks, which also may take significant time and may pose a problem for scaling of an existing Fabric network.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC introduces a notion of a **ledger snapshot for OSNs** (ordering service nodes). With this notion two new logical operations appear at the orderer REST API and in the CLI tool *osnadmin*:
+- Create a snapshot,
+- Join channel from a snapshot.
+
+Ledger snapshot is a set of files containing all the necessary information for OSN to join a channel from a block later than a genesis block (block #0) without loss of any functionality except the ability to deliver an earlier block.
+OSN ledger snapshot functionality is designed with the same principles as the peer ledger snapshot, in particular it is a full snapshot. We do not implement an incremental snapshot capability. Each snapshot is fully independent and self-sufficient and does not rely on the content of any previous snapshot. For example even if we start an OSN from a snapshot and take a new snapshot and there is no new config block in between, we do not refer to the last config block from the previous snapshot, instead we include the copy of the same last config block in the new snapshot.
+
+## Guide: Create snapshot
+
+To create a ledger snapshot use the following REST API (block_number must be supplied as a parameter in the request body):
+
+    POST /v1/mgmt/channels/{channelID}/snapshot
+
+ or the following CLI command:
+
+    $ osnadmin snapshot submitrequest -c <channelID> -b <block_number>
+
+A process of snapshot creation will start immediately if the block number is less than current ledger height. If the block number is in future (>= ledger height) then the operation will immediately fail with a reasonable error message.
+
+The process of snapshot creation is an asynchronous one, so the success of the "snapshot submitrequest" operation does not indicate that the snapshot itself is ready. The indication of the snapshot readiness is the presence of the appropriate directory in the file system (<Config.FileLedger.snapshots.rootDir>/completed/<channelID>/<block_number>).
+
+## Guide: Join channel from a snapshot
+
+To join a channel from a snapshot use the following REST API (files from the snapshot must be placed in the body, using multipart/form-data content-type):
```

Comment:
I don't know why we're coupling this API so tightly to the snapshot encoding format.

Why not simply add additional parameters to the existing join? Instead of only a config block, we could also specify a pruned block number and the other information we need to perform the replication.

If there is ever another snapshot format, then won't we run into problems?  If we have to version our APIs whenever the version of our snapshot changes, this seems like a bad situation.

Reply:
1. We would like to avoid using the same API for both "join by join-block" and "join by snapshot", because these two operations are semantically quite different. "join by join-block" implies that all the blocks starting from block #0 will be pulled into the local ledger, while "join by snapshot" is specifically designed to avoid this. That's why we prefer to define two separate APIs for these two separate operations.
2. Forward compatibility concern - resolved here:
d17c0a01acca9efe8dd3539a2806972330fea785